tensorflow/compiler/xla/service/gpu/cudnn_fused_conv_rewriter.cc

```diff
@@ -387,11 +418,34 @@ void CaptureConvGraphRecursive(HloInstruction* instr,
   final_instr = instr;
 
   HloInstruction *op, *operand0, *operand1;
+  auto fuse_amax = [&]() -> bool {
+    HloComputation* reduce_comp = op->to_apply();
+    HloInstruction* reduce_comp_root = reduce_comp->root_instruction();
+    if (ShapeUtil::IsScalar(op->shape()) &&
+        ShapeUtil::IsScalar(op->operand(1)->shape()) &&
+        op->operand(1)->IsConstant() &&
+        op->operand(1)->literal().GetAsDouble({}) <= 0. &&
+        reduce_comp_root->opcode() == HloOpcode::kMaximum &&
+        reduce_comp_root->operand(0)->opcode() == HloOpcode::kParameter &&
+        reduce_comp_root->operand(1)->opcode() == HloOpcode::kParameter) {
+      aux_outputs.emplace_back(op);
+      graph_string.AppendOp("amax", op, {operand0});
+      return true;
+    }
+    return false;
+  };
+
+  // Copy the current state in case fusion will be unsuccessful or unfavorable.
+  GraphString init_graph_string = graph_string;
+  std::vector<HloInstruction*> init_operands = operands,
+                               init_aux_outputs = aux_outputs;
+  int linear_users = 0, nonlinear_users = 0;
   for (HloInstruction* user : instr->users()) {
```

Comment:
I wonder if this logic can be clearer if, instead of checking each user if it matches a pattern, instead we check the op itself, then make a recursive call for each user. This has several advantages. For one, we don't have to explicitly "rollback" changes to local variables like init_graph_string, but instead can implicitly pass a copy of graph_string to the recursive call by passing-by-value. Second, dealing with linear_users and non_linear users is easier as we can immediately bail when we see multiple linear or nonlinear users.

Another alternative is not using recursion at all. Instead, perhaps we can structure this function as something like


```c++
HloInstruction* current_instr = ...
while (current_instr) {
  next_instr = nullptr;
  for (HloInstruction* user : instr->users()) {
    ...
    if (Match(...)) {
      graph_string.AppendOp(...)
      next_instr = current_instr
    }
    ...
  current_instr = next_instr;
}
```

Reply:
As I understand it, we would still have two states and we would conditionally assign the returned state from the recursion to the return value of the current instance. On the other hand, having a separate loop over the users of an op after each successful match may make the general structure a bit more complex (again IIUC). If you think it’s preferable, the current implementation also allows us to split the conditions for not fusing and to end the loop early if there are more than one linear or nonlinear users. We could also have a condition to return early if there are more than two users.